### PR TITLE
Encode OpenAPI refs and remove concept of "ref-safe" names

### PR DESCRIPTION
**Pull request by nguerrera**
* Correctly URI-encode OpenAPI refs when necessary. Delete broken code that attempted to address this with name mangling.

* Remove all regexes from how we decide what to inline and what to ref, and how to name things. The reasoning is now done on the type objects, and not on their serialized names.

* The current choices are preserved modulo obvious bugs, but I will open an issue to revisit some of them. It will be easier to change after this refactoring. (Edit: filed https://github.com/microsoft/cadl/issues/464)

* Centralize the logic in the shared openapi lib for reuse by cadl-autorest.

* Fix issues with the regex-based stripping of Cadl and service namespaces by replacing that with a namespace filter callback on Checker.getTypeName.

* Defend against parameter key and type name collisions with diagnostics.

### CODE REVIEW

**Comment by nguerrera** [comment]
cc @xirzec

File: packages/compiler/core/checker.ts [hunk 1]
```diff
@@ -107,8 +111,8 @@ export interface Checker {
   getLiteralType(node: NumericLiteralNode): NumericLiteralType;
   getLiteralType(node: BooleanLiteralNode): BooleanLiteralType;
   getLiteralType(node: LiteralNode): LiteralType;
-  getTypeName(type: Type): string;
-  getNamespaceString(type: NamespaceType | undefined): string;
+  getTypeName(type: Type, options?: TypeNameOptions): string;
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:114]: Not really in the scope of this PR but couldn't `getTypeName` be moved out of the checker. Doesn't feel like they depend on it.

File: packages/openapi/src/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  program: Program,
+  type: Type,
+  options: TypeNameOptions,
+  existing?: Record<string, any>
+): string {
+  const name =
+    getFriendlyName(program, type, options) ?? program.checker!.getTypeName(type, options);
+
+  if (existing && existing[name] !== undefined) {
+    reportDiagnostic(program, {
```
Comment (timotheeguerin) [inline packages/openapi/src/helpers.ts:58]: Should we report both instances conflicting. That's what we have for duplicate-symbol or duplicate routes.

Comment (timotheeguerin) [inline packages/openapi/src/helpers.ts:58]: Should we report both instances conflicting. That's what we have for duplicate-symbol or duplicate routes.

Comment (nguerrera) [reply packages/openapi/src/helpers.ts:58]: We weren't handling the collisions at all before so this is a strict improvement already.

**Review by daviwil — APPROVED** [review]
Overall, I like it!  It is a little weird to see `Rest.Resource.Something` instead of `Cadl.Rest.Resource.Something` in some places, but it's not a huge issue.

**Comment by nguerrera** [comment]
> It is a little weird to see Rest.Resource.Something instead of Cadl.Rest.Resource.Something in some places, but it's not a huge issue.

It should be the case that you don't see Cadl.Rest in the OpenAPI anymore anywhere. It used to be rather haphazard when we'd remove Cadl. Now we remove it thoroughly and some things look a little weird as a result. I do have a note on #464 to revisit this.

**Pull request merged**